File: codes/scala/chapter_searching/linear_search.scala


/**
 * File: linear_search.scala
 * Author: WangBlue (wangjiping596@gmail.com)
 * Create: 2024-08-15 10:44
 */

class linear_search {
  /* 线性查找（数组） */
  def linearSearchArray(nums: Array[Int], target: Int): Int = {
    // 遍历数组
    for (num <- nums) {
      // 找到目标元素，返回其索引
      if (num == target)
        return nums.indexOf(num)
    }
    // 未找到目标值，返回-1
    -1
  }

  /* 线性查找（链表） */
  def linearSearchLinkedList(head: ListNode, target: Int): ListNode = {
    // 遍历链表
    var curr = head
    while (curr != null) {
      // 找到目标值，返回节点
      if (curr.value == target)
        return curr
      // 未找到目标值，移动到下一个节点
      curr = curr.next
    }
    // 未找到目标值，返回null
    null
  }
}

object linear_search {
  def main(args: Array[String]): Unit = {
    val nums = Array(1, 5, 3, 2, 4, 7, 5, 9, 10, 8)
    val target = 3
    /* 在数组中执行线性查找 */

    val linearSearch = new linear_search()
    val index = linearSearch.linearSearchArray(nums, target)
    println("目标元素 3 的索引 = " + index)

    /* 在链表中执行线性查找 *//* 在链表中执行线性查找 */
    val head = new ListNode().arrToLinkedList(nums)
    val node = linearSearch.linearSearchLinkedList(head, target)
    println("目标节点值 3 的对应节点对象为 " + node.value)
  }
}

File: codes/rust/chapter_searching/linear_search.rs
/*
 * File: linear_search.rs
 * Created Time: 2023-07-09
 * Author: codingonion (coderonion@gmail.com)
 */

use hello_algo_rust::linked_list::{LinkedList, ListLink, ListNode};
use std::cell::RefCell;
use std::rc::Rc;

/* 线性查找（数组） */
pub fn linear_search_array(nums: &[i32], target: i32) -> Option<usize> {
    // 遍历数组
    for (i, &num) in nums.iter().enumerate() {
        // 找到目标元素，返回其索引
        if num == target {
            return Some(i);
        }
    }
    // 未找到目标元素，返回 None
    None
}

/* 线性查找（链表） */
pub fn linear_search_linked_list(
    head: &Rc<RefCell<ListNode<i32>>>,
    target: i32,
) -> Option<Rc<RefCell<ListNode<i32>>>> {
    // 找到目标节点，返回之
    if head.borrow().val == target {
        return Some(Rc::clone(head));
    };
    // 找到目标节点，返回之
    if let Some(node) = &head.borrow().next {
        return linear_search_linked_list(node, target);
    }
    // 未找到目标节点，返回 None
    None
}

/* Driver Code */
fn main() {
    let target = 3;

    /* 在数组中执行线性查找 */
    let nums = [1, 5, 3, 2, 4, 7, 5, 9, 10, 8];
    let index = linear_search_array(&nums, target);
    println!("目标元素 3 的索引 = {index:?}");

    /* 在链表中执行线性查找 */
    let head = ListLink::try_from_array(nums).unwrap();
    let node = linear_search_linked_list(&head, target);
    println!("目标节点值 3 的对应节点对象为 {node:?}");
}
